File: 백준/Gold/25332. 수들의 합 8/수들의 합 8.py
import sys
from collections import defaultdict

input = sys.stdin.readline


n = int(input())
aList = [0] + list(map(int, input().split()))
bList = [0] + list(map(int, input().split()))

prefixSum = [0] * (n + 1)
for i in range(1, n + 1):
    prefixSum[i] = prefixSum[i - 1] + aList[i] - bList[i]

counter = defaultdict(int)
answer = 0

for i in range(n, 0, -1):
    v = prefixSum[i]

    if v == 0:
        answer += 1

    answer += counter[v]
    counter[v] += 1

print(answer)

File: 백준/Gold/25332. 수들의 합 8/수들의 합 8.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;

public class Main {

  static StringBuilder sb = new StringBuilder();
  static int n;
  static int[] prefixSum, a, b;
  static Map<Integer, Integer> counter;

  public static void main(String[] args) throws IOException {
    setUp();

    long answer = 0L;
    for (int i = n; i > 0; i--) {
      int curr = prefixSum[i];

      if (curr == 0) {
        answer++;
      }

      answer += counter.getOrDefault(curr, 0);
      counter.put(curr, counter.getOrDefault(curr, 0) + 1);
    }

    sb.append(answer);

    output();
  }

  private static void setUp() throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    StringTokenizer st;

    n = Integer.parseInt(br.readLine());
    prefixSum = new int[n + 1];
    a = new int[n + 1];
    b = new int[n + 1];

    st = new StringTokenizer(br.readLine());
    for (int i = 1; i < n + 1; i++) {
      a[i] = Integer.parseInt(st.nextToken());
    }
    st = new StringTokenizer(br.readLine());
    for (int i = 1; i < n + 1; i++) {
      b[i] = Integer.parseInt(st.nextToken());
    }

    for (int i = 1; i < n + 1; i++) {
      prefixSum[i] = prefixSum[i - 1] + a[i] - b[i];
    }

    counter = new HashMap<>();
  }

  private static void output() throws IOException {
    BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    bw.write(sb.toString());
    bw.flush();
    bw.close();
  }

}